<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Summary" data-toc-modified-id="Summary-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Summary</a></span></li><li><span><a href="#Data-First-Look" data-toc-modified-id="Data-First-Look-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Data First Look</a></span></li><li><span><a href="#Data-Cleaning" data-toc-modified-id="Data-Cleaning-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Data Cleaning</a></span><ul class="toc-item"><li><span><a href="#Drop-Unnamed-Columns" data-toc-modified-id="Drop-Unnamed-Columns-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Drop Unnamed Columns</a></span></li><li><span><a href="#Missing-Data-Points" data-toc-modified-id="Missing-Data-Points-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Missing Data Points</a></span><ul class="toc-item"><li><span><a href="#Counts" data-toc-modified-id="Counts-3.2.1"><span class="toc-item-num">3.2.1&nbsp;&nbsp;</span>Counts</a></span></li><li><span><a href="#Percent-Missing" data-toc-modified-id="Percent-Missing-3.2.2"><span class="toc-item-num">3.2.2&nbsp;&nbsp;</span>Percent Missing</a></span></li><li><span><a href="#Replace-NaNs" data-toc-modified-id="Replace-NaNs-3.2.3"><span class="toc-item-num">3.2.3&nbsp;&nbsp;</span>Replace NaNs</a></span></li></ul></li><li><span><a href="#APR-to-Float" data-toc-modified-id="APR-to-Float-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>APR to Float</a></span></li><li><span><a href="#Test-Group-to-Bool" data-toc-modified-id="Test-Group-to-Bool-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Test Group to Bool</a></span></li></ul></li><li><span><a href="#Univariate-Exploration" data-toc-modified-id="Univariate-Exploration-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Univariate Exploration</a></span><ul class="toc-item"><li><span><a href="#FICO-Score-Application-Counts" data-toc-modified-id="FICO-Score-Application-Counts-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>FICO Score Application Counts</a></span></li><li><span><a href="#APR-Application-Counts" data-toc-modified-id="APR-Application-Counts-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>APR Application Counts</a></span></li></ul></li><li><span><a href="#Bivariate-Exploration" data-toc-modified-id="Bivariate-Exploration-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Bivariate Exploration</a></span><ul class="toc-item"><li><span><a href="#FICO-Scores-and-Test-Group" data-toc-modified-id="FICO-Scores-and-Test-Group-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>FICO Scores and Test Group</a></span></li><li><span><a href="#APR-per-Sale-for-Control-Group" data-toc-modified-id="APR-per-Sale-for-Control-Group-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>APR per Sale for Control Group</a></span></li><li><span><a href="#FICO-and-APR-for-Test-and-Control" data-toc-modified-id="FICO-and-APR-for-Test-and-Control-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>FICO and APR for Test and Control</a></span></li></ul></li><li><span><a href="#Question-Answers-and-Analysis" data-toc-modified-id="Question-Answers-and-Analysis-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Question Answers and Analysis</a></span><ul class="toc-item"><li><span><a href="#Q1A:-What’s-the-current-APR-pricing-strategy-for-the-different-FICO-buckets?" data-toc-modified-id="Q1A:-What’s-the-current-APR-pricing-strategy-for-the-different-FICO-buckets?-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Q1A: What’s the current APR pricing strategy for the different FICO buckets?</a></span></li><li><span><a href="#Q1B:-How-is-the-randomized-test-structured?" data-toc-modified-id="Q1B:-How-is-the-randomized-test-structured?-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>Q1B: How is the randomized test structured?</a></span></li><li><span><a href="#Q1C:-How-do-APR's-vary-by-test-group?" data-toc-modified-id="Q1C:-How-do-APR's-vary-by-test-group?-6.3"><span class="toc-item-num">6.3&nbsp;&nbsp;</span>Q1C: How do APR's vary by test group?</a></span></li><li><span><a href="#Q2A:-How-many-applications-is-Carvana-able-to-convert-to-a-sale?" data-toc-modified-id="Q2A:-How-many-applications-is-Carvana-able-to-convert-to-a-sale?-6.4"><span class="toc-item-num">6.4&nbsp;&nbsp;</span>Q2A: How many applications is Carvana able to convert to a sale?</a></span></li><li><span><a href="#Q2B:-How-does-that-vary-by-FICO?" data-toc-modified-id="Q2B:-How-does-that-vary-by-FICO?-6.5"><span class="toc-item-num">6.5&nbsp;&nbsp;</span>Q2B: How does that vary by FICO?</a></span></li><li><span><a href="#Q2C:-Using-the-APR-test-group-data,-how-does-the-APR-affect-conversion?" data-toc-modified-id="Q2C:-Using-the-APR-test-group-data,-how-does-the-APR-affect-conversion?-6.6"><span class="toc-item-num">6.6&nbsp;&nbsp;</span>Q2C: Using the APR test group data, how does the APR affect conversion?</a></span></li><li><span><a href="#Q3A:-What-does-Carvana-make-in-margin-overall,-per-sale-and-per-application?" data-toc-modified-id="Q3A:-What-does-Carvana-make-in-margin-overall,-per-sale-and-per-application?-6.7"><span class="toc-item-num">6.7&nbsp;&nbsp;</span>Q3A: What does Carvana make in margin overall, per sale and per application?</a></span></li><li><span><a href="#Q3B:-How-does-APR-affect-the-finance-margin-per-sale?" data-toc-modified-id="Q3B:-How-does-APR-affect-the-finance-margin-per-sale?-6.8"><span class="toc-item-num">6.8&nbsp;&nbsp;</span>Q3B: How does APR affect the finance margin per sale?</a></span></li><li><span><a href="#Q3C:-How-does-FICO-affect-the-finance-margin-per-sale?" data-toc-modified-id="Q3C:-How-does-FICO-affect-the-finance-margin-per-sale?-6.9"><span class="toc-item-num">6.9&nbsp;&nbsp;</span>Q3C: How does FICO affect the finance margin per sale?</a></span></li><li><span><a href="#Q4:-If-Carvana-wants-to-optimize-total-margin,-do-you-think-Carvana-should-raise-or-lower-rates-on-700-FICO-customers?--What-about-600-FICO-customers?" data-toc-modified-id="Q4:-If-Carvana-wants-to-optimize-total-margin,-do-you-think-Carvana-should-raise-or-lower-rates-on-700-FICO-customers?--What-about-600-FICO-customers?-6.10"><span class="toc-item-num">6.10&nbsp;&nbsp;</span>Q4: If Carvana wants to optimize total margin, do you think Carvana should raise or lower rates on 700 FICO customers?  What about 600 FICO customers?</a></span></li><li><span><a href="#Q5:-5.----Other-than-these-cut-and-dry-numbers,-can-you-think-of-other-factors-Carvana-might-want-to-take-into-account-before-changing-rates?" data-toc-modified-id="Q5:-5.----Other-than-these-cut-and-dry-numbers,-can-you-think-of-other-factors-Carvana-might-want-to-take-into-account-before-changing-rates?-6.11"><span class="toc-item-num">6.11&nbsp;&nbsp;</span>Q5: 5.    Other than these cut and dry numbers, can you think of other factors Carvana might want to take into account before changing rates?</a></span></li></ul></li></ul></div>

# Summary

Questions have been divided into sub-sections. Each question is answered in the section and is followed by the supporting analysis and tables.

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.plotly as py
import plotly.graph_objs as go
from plotly import tools

In [2]:
df=pd.read_csv('APR_strategy_w._randomized_experiment.csv')

# Data First Look

In [3]:
df.shape

(20000, 11)

In [4]:
df.head()

app_id  fico apr_test_group   apr  converted to sale  vehicle margin  \
0       1   700        control  4.0%              False             NaN   
1       2   700           test  3.0%              False             NaN   
2       3   700           test  3.0%              False             NaN   
3       4   700        control  4.0%              False             NaN   
4       5   700        control  4.0%               True          1000.0   

   finance margin  Unnamed: 7  Unnamed: 8  Unnamed: 9  Unnamed: 10  
0             NaN         NaN         NaN         NaN          NaN  
1             NaN         NaN         NaN         NaN          NaN  
2             NaN         NaN         NaN         NaN          NaN  
3             NaN         NaN         NaN         NaN          NaN  
4           600.0         NaN         NaN         NaN          NaN

In [5]:
df.describe()

app_id         fico  vehicle margin  finance margin  Unnamed: 7  \
count  20000.000000  20000.00000          5021.0     5021.000000         0.0   
mean   10000.500000    649.99500          1000.0      488.777136         NaN   
std     5773.647028     50.00125             0.0      127.370035         NaN   
min        1.000000    600.00000          1000.0      350.000000         NaN   
25%     5000.750000    600.00000          1000.0      350.000000         NaN   
50%    10000.500000    600.00000          1000.0      400.000000         NaN   
75%    15000.250000    700.00000          1000.0      600.000000         NaN   
max    20000.000000    700.00000          1000.0      650.000000         NaN   

       Unnamed: 8  Unnamed: 9  Unnamed: 10  
count         0.0         0.0          0.0  
mean          NaN         NaN          NaN  
std           NaN         NaN          NaN  
min           NaN         NaN          NaN  
25%           NaN         NaN          NaN  
50%           NaN         NaN          NaN  
75%           NaN         NaN          NaN  
max           NaN         NaN          NaN

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 11 columns):
app_id               20000 non-null int64
fico                 20000 non-null int64
apr_test_group       20000 non-null object
apr                  20000 non-null object
converted to sale    20000 non-null bool
vehicle margin       5021 non-null float64
finance margin       5021 non-null float64
Unnamed: 7           0 non-null float64
Unnamed: 8           0 non-null float64
Unnamed: 9           0 non-null float64
Unnamed: 10          0 non-null float64
dtypes: bool(1), float64(6), int64(2), object(2)
memory usage: 1.5+ MB


# Data Cleaning

In [7]:
df_clean = df.copy()

## Drop Unnamed Columns

In [8]:
df_drop = df.copy()

In [9]:
df_drop = df_drop.loc[:, ~df_drop.columns.str.contains('^Unnamed')]

In [10]:
df_clean = df_drop.copy()
df_clean.head()

app_id  fico apr_test_group   apr  converted to sale  vehicle margin  \
0       1   700        control  4.0%              False             NaN   
1       2   700           test  3.0%              False             NaN   
2       3   700           test  3.0%              False             NaN   
3       4   700        control  4.0%              False             NaN   
4       5   700        control  4.0%               True          1000.0   

   finance margin  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4           600.0

## Missing Data Points

In [11]:
df_missing = df_clean.copy()

### Counts

In [12]:
missing_count = df_missing.isnull().sum()
print (missing_count)

app_id                   0
fico                     0
apr_test_group           0
apr                      0
converted to sale        0
vehicle margin       14979
finance margin       14979
dtype: int64


### Percent Missing

In [13]:
total_cells = np.product(df_missing.shape)
total_missing = missing_count.sum()
(total_missing/total_cells) * 100

21.39857142857143

### Replace NaNs

In [14]:
df_missing = df_missing.fillna(0)

In [15]:
df_clean = df_missing.copy()
df_clean.head()

app_id  fico apr_test_group   apr  converted to sale  vehicle margin  \
0       1   700        control  4.0%              False             0.0   
1       2   700           test  3.0%              False             0.0   
2       3   700           test  3.0%              False             0.0   
3       4   700        control  4.0%              False             0.0   
4       5   700        control  4.0%               True          1000.0   

   finance margin  
0             0.0  
1             0.0  
2             0.0  
3             0.0  
4           600.0

## APR to Float

In [19]:
df_apr = df_clean.copy()

In [20]:
df_apr['apr'] = df_apr['apr'].str.replace('%','').astype(float)

In [21]:
df_clean = df_apr.copy()
df_clean.head()

app_id  fico apr_test_group  apr  converted to sale  vehicle margin  \
0       1   700        control  4.0              False             0.0   
1       2   700           test  3.0              False             0.0   
2       3   700           test  3.0              False             0.0   
3       4   700        control  4.0              False             0.0   
4       5   700        control  4.0               True          1000.0   

   finance margin  
0             0.0  
1             0.0  
2             0.0  
3             0.0  
4           600.0

## Test Group to Bool

In [22]:
#df_test_group = df_clean.copy()

In [23]:
#df_test_group['apr_test_group'] = np.where(df_test_group['apr_test_group'].str.contains('test'),1,0)

In [24]:
#df_clean = df_test_group.copy()
#df_clean.head()

# Univariate Exploration

In [25]:
df_analysis = df_clean.copy()

## FICO Score Application Counts

In [131]:
def fico_counts():
    
    fico_600 = len(df_analysis[(df_analysis['fico']==600)].index)
    fico_700 = len(df_analysis[(df_analysis['fico']==700)].index)
    
    data = [go.Bar(
                x=['600', '700'],
                y=[fico_600, fico_700]
    )]
        
    layout = go.Layout(
        title = 'FICO Score Counts',
        xaxis=dict(
            title='FICO Scores',
            tickvals = [600,700]
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = fico_counts()
py.iplot(fig)

## APR Application Counts

In [134]:
def apr_counts():
    
    apr_30 = len(df_analysis[(df_analysis['apr']==3.0)].index)
    apr_40 = len(df_analysis[(df_analysis['apr']==4.0)].index)
    apr_44 = len(df_analysis[(df_analysis['apr']==4.4)].index)
    apr_54 = len(df_analysis[(df_analysis['apr']==5.4)].index)
    
    data = [go.Bar(
                x=['3.0', '4.0', '4.4', '5.4'],
                y=[apr_30, apr_40, apr_44, apr_54]
    )]
        
    layout = go.Layout(
        title = 'APR Counts',
        xaxis=dict(
            title='ACRs',
            tickvals = [3.0,4.0,4.4,5.4]
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = apr_counts()
py.iplot(fig)

# Bivariate Exploration

## FICO Scores and Test Group

In [56]:
 def fico_test():
        
    test_fico600 = len(df_analysis[(df_analysis['fico']==600) & (df_analysis['apr_test_group']=='test')].index)
    test_fico700 = len(df_analysis[(df_analysis['fico']==700) & (df_analysis['apr_test_group']=='test')].index)
    ctrl_fico600 = len(df_analysis[(df_analysis['fico']==600) & (df_analysis['apr_test_group']=='control')].index)
    ctrl_fico700 = len(df_analysis[(df_analysis['fico']==700) & (df_analysis['apr_test_group']=='control')].index)
    
    trace1 = go.Bar(
        x=['Test', 'Control'],
        y=[test_fico600, ctrl_fico600],
        name='600 FICO'
    )

    trace2 = go.Bar(
        x=['Test', 'Control'],
        y=[test_fico700, ctrl_fico700],
        name='700 FICO'
    )
    
    data = [trace1, trace2]
    layout = go.Layout(
        barmode='group',
        title = 'FICO Scores per Test Group',
        xaxis=dict(
            title='Test Groups'
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = fico_test()
py.iplot(fig)

## APR per Sale for Control Group

In [110]:
def plot_apr_sale_ctrl():
    test_sold_apr30 = len(df_analysis[(df_analysis['apr']==3.0) 
                        &(df_analysis['apr_test_group']=='control')
                        &(df_analysis['converted to sale']==True)].index)
    test_sold_apr40 = len(df_analysis[(df_analysis['apr']==4.0) 
                        &(df_analysis['apr_test_group']=='control')
                        &(df_analysis['converted to sale']==True)].index)
    test_sold_apr44 = len(df_analysis[(df_analysis['apr']==4.4) 
                        &(df_analysis['apr_test_group']=='control')
                        &(df_analysis['converted to sale']==True)].index)
    test_sold_apr54 = len(df_analysis[(df_analysis['apr']==5.4)
                        & (df_analysis['apr_test_group']=='control')
                        &(df_analysis['converted to sale']==True)].index)
    test_nosale_apr30 = len(df_analysis[(df_analysis['apr']==3.0) 
                        &(df_analysis['apr_test_group']=='control')
                        &(df_analysis['converted to sale']==False)].index)
    test_nosale_apr40 = len(df_analysis[(df_analysis['apr']==4.0) 
                        &(df_analysis['apr_test_group']=='control')
                        &(df_analysis['converted to sale']==False)].index)
    test_nosale_apr44 = len(df_analysis[(df_analysis['apr']==4.4) 
                        &(df_analysis['apr_test_group']=='control')
                        &(df_analysis['converted to sale']==False)].index)
    test_nosale_apr54 = len(df_analysis[(df_analysis['apr']==5.4)
                        & (df_analysis['apr_test_group']=='control')
                        &(df_analysis['converted to sale']==False)].index)
   
    trace1 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[test_sold_apr30, test_nosale_apr30],
        name='3.0 APR'
    )
    trace2 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[test_sold_apr40, test_nosale_apr40],
        name='4.0 APR'
    )

    trace3 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[test_sold_apr44, test_nosale_apr44],
        name='4.4 APR'
    )

    trace4 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[test_sold_apr54, test_nosale_apr54],
        name='5.4 APR'
    )

    data = [trace1, trace2, trace3, trace4]
    layout = go.Layout(
        barmode='group',
        title = 'APR per Sale for Control Group',
        xaxis=dict(
            title='Test Groups'
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = plot_apr_sale_ctrl()
py.iplot(fig)  

## FICO and APR for Test and Control

In [111]:

def plot_apr_test():
    test_apr30 = len(df_analysis[(df_analysis['apr']==3.0) & (df_analysis['apr_test_group']=='test')].index)
    test_apr40 = len(df_analysis[(df_analysis['apr']==4.0) & (df_analysis['apr_test_group']=='test')].index)
    test_apr44 = len(df_analysis[(df_analysis['apr']==4.4) & (df_analysis['apr_test_group']=='test')].index)
    test_apr54 = len(df_analysis[(df_analysis['apr']==5.4) & (df_analysis['apr_test_group']=='test')].index)

    ctrl_apr30 = len(df_analysis[(df_analysis['apr']==3.0) & (df_analysis['apr_test_group']=='control')].index)
    ctrl_apr40 = len(df_analysis[(df_analysis['apr']==4.0) & (df_analysis['apr_test_group']=='control')].index)
    ctrl_apr44 = len(df_analysis[(df_analysis['apr']==4.4) & (df_analysis['apr_test_group']=='control')].index)
    ctrl_apr54 = len(df_analysis[(df_analysis['apr']==5.4) & (df_analysis['apr_test_group']=='control')].index)

    test_fico600 = len(df_analysis[(df_analysis['fico']==600) & (df_analysis['apr_test_group']=='test')].index)
    test_fico700 = len(df_analysis[(df_analysis['fico']==700) & (df_analysis['apr_test_group']=='test')].index)
    ctrl_fico600 = len(df_analysis[(df_analysis['fico']==600) & (df_analysis['apr_test_group']=='control')].index)
    ctrl_fico700 = len(df_analysis[(df_analysis['fico']==700) & (df_analysis['apr_test_group']=='control')].index)
    
    # APR
    trace1 = go.Bar(
        x=['Test', 'Control'],
        y=[test_apr30, ctrl_apr30],
        name='3.0 APR'
    )
    trace2 = go.Bar(
        x=['Test', 'Control'],
        y=[test_apr40, ctrl_apr40],
        name='4.0 APR'
    )

    trace3 = go.Bar(
        x=['Test', 'Control'],
        y=[test_apr44, ctrl_apr44],
        name='4.4 APR'
    )

    trace4 = go.Bar(
        x=['Test', 'Control'],
        y=[test_apr54, ctrl_apr54],
        name='5.4 APR'
    )
    
    # FICO
    trace5 = go.Bar(
        x=['Test', 'Control'],
        y=[test_fico600, ctrl_fico600],
        name='600 FICO'
    )

    trace6 = go.Bar(
        x=['Test', 'Control'],
        y=[test_fico700, ctrl_fico700],
        name='700 FICO'
    )
  
    fig = tools.make_subplots(rows=1, cols=2, 
                              subplot_titles=('APR',
                                            'FICO'))
    
    fig.append_trace(trace1, 1,1)
    fig.append_trace(trace2, 1,1)
    fig.append_trace(trace3, 1,1)
    fig.append_trace(trace4, 1,1)
    fig.append_trace(trace5, 1,2)
    fig.append_trace(trace6, 1,2)
    
    fig['layout'].update(height=500, width=700)
    fig['layout']['xaxis1'].update(title='APR Scores')
    fig['layout']['yaxis1'].update(title='Applications')
    fig['layout']['xaxis2'].update(title='FICO Scores')
    
    return (fig)

fig = plot_apr_test()
py.iplot(fig)

This is the format of your plot grid:
[ (1,1) x1,y1 ]  [ (1,2) x2,y2 ]



# Question Answers and Analysis

## Q1A: What’s the current APR pricing strategy for the different FICO buckets?
A better FICO score means a lower APR.

An APR of 3.0 or 4.0 are used for applications with a FICO score of 700 with slightly more applications having a APR of 3.0. Applications with a 700 FICO score have no APRs of 4.4 or 5.4. 

An APR of 4.4 or 5.4 are used for applications with a FICO score of 600 with slightly more applications having a APR of 5.4. Applications with a 600 FICO score have no APRs of 3.0 or 4.0.

In [31]:
def plot_apr_fico():
    f600_apr30 = len(df_analysis[(df_analysis['apr']==3.0) & (df_analysis['fico']==600)].index)
    f600_apr40 = len(df_analysis[(df_analysis['apr']==4.0) & (df_analysis['fico']==600)].index)
    f600_apr44 = len(df_analysis[(df_analysis['apr']==4.4) & (df_analysis['fico']==600)].index)
    f600_apr54 = len(df_analysis[(df_analysis['apr']==5.4) & (df_analysis['fico']==600)].index)

    f700_apr30 = len(df_analysis[(df_analysis['apr']==3.0) & (df_analysis['fico']==700)].index)
    f700_apr40 = len(df_analysis[(df_analysis['apr']==4.0) & (df_analysis['fico']==700)].index)
    f700_apr44 = len(df_analysis[(df_analysis['apr']==4.4) & (df_analysis['fico']==700)].index)
    f700_apr54 = len(df_analysis[(df_analysis['apr']==5.4) & (df_analysis['fico']==700)].index)


    trace1 = go.Bar(
        x=['600', '700'],
        y=[f600_apr30, f700_apr30],
        name='3.0 APR'
    )
    trace2 = go.Bar(
        x=['600', '700'],
        y=[f600_apr40, f700_apr40],
        name='4.0 APR'
    )

    trace3 = go.Bar(
        x=['600', '700'],
        y=[f600_apr44,f700_apr44],
        name='4.4 APR'
    )

    trace4 = go.Bar(
        x=['600', '700'],
        y=[f600_apr54, f700_apr54],
        name='5.4 APR'
    )

    data = [trace1, trace2, trace3, trace4]
    layout = go.Layout(
        barmode='group',
        title = 'APR per FICO Bucket',
        xaxis=dict(
            title='FICO Buckets'
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = plot_apr_fico()
py.iplot(fig)

## Q1B: How is the randomized test structured? 

40k applications have been divided evenly into a control group and test group.

In [50]:
def plot_test():
    test = len((df_analysis['apr_test_group']=='test').index)

    ctrl = len((df_analysis['apr_test_group']=='control').index)
    
    data = [go.Bar(
                x=['Test', 'Control'],
                y=[test, ctrl]
    )]
        
    layout = go.Layout(
        title = 'Test Structure',
        xaxis=dict(
            title='Test Groups'
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = plot_test()
py.iplot(fig)

## Q1C: How do APR's vary by test group?
The test group is has APRs of 3.0 and 4.4 where the control group has APRs of 4.0 and 5.4.

In [57]:
def plot_apr_fico():
    test_apr30 = len(df_analysis[(df_analysis['apr']==3.0) & (df_analysis['apr_test_group']=='test')].index)
    test_apr40 = len(df_analysis[(df_analysis['apr']==4.0) & (df_analysis['apr_test_group']=='test')].index)
    test_apr44 = len(df_analysis[(df_analysis['apr']==4.4) & (df_analysis['apr_test_group']=='test')].index)
    test_apr54 = len(df_analysis[(df_analysis['apr']==5.4) & (df_analysis['apr_test_group']=='test')].index)

    ctrl_apr30 = len(df_analysis[(df_analysis['apr']==3.0) & (df_analysis['apr_test_group']=='control')].index)
    ctrl_apr40 = len(df_analysis[(df_analysis['apr']==4.0) & (df_analysis['apr_test_group']=='control')].index)
    ctrl_apr44 = len(df_analysis[(df_analysis['apr']==4.4) & (df_analysis['apr_test_group']=='control')].index)
    ctrl_apr54 = len(df_analysis[(df_analysis['apr']==5.4) & (df_analysis['apr_test_group']=='control')].index)


    trace1 = go.Bar(
        x=['Test', 'Control'],
        y=[test_apr30, ctrl_apr30],
        name='3.0 APR'
    )
    trace2 = go.Bar(
        x=['Test', 'Control'],
        y=[test_apr40, ctrl_apr40],
        name='4.0 APR'
    )

    trace3 = go.Bar(
        x=['Test', 'Control'],
        y=[test_apr44, ctrl_apr44],
        name='4.4 APR'
    )

    trace4 = go.Bar(
        x=['Test', 'Control'],
        y=[test_apr54, ctrl_apr54],
        name='5.4 APR'
    )

    data = [trace1, trace2, trace3, trace4]
    layout = go.Layout(
        barmode='group',
        title = 'APR per Test Group',
        xaxis=dict(
            title='Test Groups'
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = plot_apr_fico()
py.iplot(fig)

## Q2A: How many applications is Carvana able to convert to a sale? 

20k (50%) of applications are converted to a sale.

In [60]:
def plot_sales():
    test = len((df_analysis['converted to sale']==True).index)

    ctrl = len((df_analysis['converted to sale']==False).index)
    
    data = [go.Bar(
                x=['Sold', 'Not Sold'],
                y=[test, ctrl]
    )]
        
    layout = go.Layout(
        title = 'Sales',
        xaxis=dict(
            title='Sale Status'
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = plot_sales()
py.iplot(fig)

## Q2B: How does that vary by FICO?

589 more applications with a FICO score of 700 are sold than applications with a FICO score of 600. 

In [61]:
 def fico_test():
        
    sale_fico600 = len(df_analysis[(df_analysis['fico']==600) & (df_analysis['converted to sale']==True)].index)
    sale_fico700 = len(df_analysis[(df_analysis['fico']==700) & (df_analysis['converted to sale']==True)].index)
    no_sale_fico600 = len(df_analysis[(df_analysis['fico']==600) & (df_analysis['converted to sale']==False)].index)
    no_sale_fico700 = len(df_analysis[(df_analysis['fico']==700) & (df_analysis['converted to sale']==False)].index)
    
    trace1 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[sale_fico600, no_sale_fico600],
        name='600 FICO'
    )

    trace2 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[sale_fico700, no_sale_fico700],
        name='700 FICO'
    )
    
    data = [trace1, trace2]
    layout = go.Layout(
        barmode='group',
        title = 'FICO Scores per Test Group',
        xaxis=dict(
            title='Test Groups'
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = fico_test()
py.iplot(fig)

## Q2C: Using the APR test group data, how does the APR affect conversion?

For the test group, the majority of applicatons with 3.0 and 4.4 APR did not result in a sale (2677 sold vs 7302 not sold).

333 more sales were made with an APR of 3.0 than 4.4 (1505 vs 1172).

In [146]:
def plot_apr_sale_test():
    test_sold_apr30 = len(df_analysis[(df_analysis['apr']==3.0) 
                        &(df_analysis['apr_test_group']=='test')
                        &(df_analysis['converted to sale']==True)].index)
    test_sold_apr40 = len(df_analysis[(df_analysis['apr']==4.0) 
                        &(df_analysis['apr_test_group']=='test')
                        &(df_analysis['converted to sale']==True)].index)
    test_sold_apr44 = len(df_analysis[(df_analysis['apr']==4.4) 
                        &(df_analysis['apr_test_group']=='test')
                        &(df_analysis['converted to sale']==True)].index)
    test_sold_apr54 = len(df_analysis[(df_analysis['apr']==5.4)
                        & (df_analysis['apr_test_group']=='test')
                        &(df_analysis['converted to sale']==True)].index)
    test_nosale_apr30 = len(df_analysis[(df_analysis['apr']==3.0) 
                        &(df_analysis['apr_test_group']=='test')
                        &(df_analysis['converted to sale']==False)].index)
    test_nosale_apr40 = len(df_analysis[(df_analysis['apr']==4.0) 
                        &(df_analysis['apr_test_group']=='test')
                        &(df_analysis['converted to sale']==False)].index)
    test_nosale_apr44 = len(df_analysis[(df_analysis['apr']==4.4) 
                        &(df_analysis['apr_test_group']=='test')
                        &(df_analysis['converted to sale']==False)].index)
    test_nosale_apr54 = len(df_analysis[(df_analysis['apr']==5.4)
                        & (df_analysis['apr_test_group']=='test')
                        &(df_analysis['converted to sale']==False)].index)
   
    trace1 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[test_sold_apr30, test_nosale_apr30],
        name='3.0 APR'
    )
    trace2 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[test_sold_apr40, test_nosale_apr40],
        name='4.0 APR'
    )

    trace3 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[test_sold_apr44, test_nosale_apr44],
        name='4.4 APR'
    )

    trace4 = go.Bar(
        x=['Sold', 'Not Sold'],
        y=[test_sold_apr54, test_nosale_apr54],
        name='5.4 APR'
    )

    data = [trace1, trace2, trace3, trace4]
    layout = go.Layout(
        barmode='group',
        title = 'APR per Sale for Test Group',
        xaxis=dict(
            title='Test Groups'
        ),
        yaxis=dict(
            title='Applications'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = plot_apr_sale_test()
py.iplot(fig)    

## Q3A: What does Carvana make in margin overall, per sale and per application?

Overall: $7,475,150.00

Per Sale: $1,488.77

Per Application: $373.75

In [79]:
df_margin = df_analysis.copy()
df_margin['margin'] = df_margin['vehicle margin'] +   df_margin['finance margin']
df_margin.head()

app_id  fico apr_test_group  apr  converted to sale  vehicle margin  \
0       1   700        control  4.0              False             0.0   
1       2   700           test  3.0              False             0.0   
2       3   700           test  3.0              False             0.0   
3       4   700        control  4.0              False             0.0   
4       5   700        control  4.0               True          1000.0   

   finance margin  margin  
0             0.0     0.0  
1             0.0     0.0  
2             0.0     0.0  
3             0.0     0.0  
4           600.0  1600.0

In [78]:
margin_overall = df_margin['margin'].sum()
print (margin_overall)

7475150.0


In [80]:
margin_per_sale = df_margin[df_margin['converted to sale']==True]['margin'].mean()
print (margin_per_sale)

1488.7771360286795


In [81]:
margin_per_application = df_margin['margin'].mean()
print (margin_per_application)

373.7575


## Q3B: How does APR affect the finance margin per sale? 

The finance margin is highest for highest APR of 5.4 (\$650) and lowest for the lowest APR of 3.0 (\$350.00).  However, that is reversed for the middle APRs. The finance margin for an APR of 4 is \$600 (only \$50 less an APR of 5.4). The finance margin for an APR of 4.4 is \$400 (only \$50 more than an APR of 3.0).

In [147]:
def apr_fmargin_sale():
    
    margin_30 = df_margin[(df_margin['apr']==3.0) & (df_margin['converted to sale']==True)]['finance margin'].mean()
    margin_40 = df_margin[(df_margin['apr']==4.0) & (df_margin['converted to sale']==True)]['finance margin'].mean()
    margin_44 = df_margin[(df_margin['apr']==4.4) & (df_margin['converted to sale']==True)]['finance margin'].mean()
    margin_54 = df_margin[(df_margin['apr']==5.4) & (df_margin['converted to sale']==True)]['finance margin'].mean()
    
    data = [go.Bar(
                x=['3.0', '4.0', '4.4', '5.4'],
                y=[margin_30, margin_40, margin_44, margin_54]
    )]
        
    layout = go.Layout(
        title = 'APR Effect on Finance Margin per Sale',
        xaxis=dict(
            title='APR',
            tickvals = [3.0,4.0,4.4,5.4]
        ),
        yaxis=dict(
            title='Mean Finance Margin'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = apr_fmargin_sale()
py.iplot(fig)

## Q3C: How does FICO affect the finance margin per sale? 

Applicaitons with a FICO score of 600 have a finance margin of $51.92  per sale more than applicaitons with a FICO score of 700.

In [100]:
def fico_fmargin_sale():
    
    fico_600 = df_margin[(df_margin['fico']==600) & (df_margin['converted to sale']==True)]['finance margin'].mean()
    fico_700 = df_margin[(df_margin['fico']==700) & (df_margin['converted to sale']==True)]['finance margin'].mean()
    
    data = [go.Bar(
                x=['600', '700'],
                y=[fico_600, fico_700]
    )]
        
    layout = go.Layout(
        title = 'FICO Effect on Finance Margin per Sale',
        xaxis=dict(
            title='FICO',
            tickvals = [600,700]
        ),
        yaxis=dict(
            title='Mean Finance Margin'
        )
    )

    fig = go.Figure(data=data, layout=layout)
    return(fig)

fig = fico_fmargin_sale()
py.iplot(fig)

## Q4: If Carvana wants to optimize total margin, do you think Carvana should raise or lower rates on 700 FICO customers?  What about 600 FICO customers?

An increase in rates for 600 and 700 FICO customer would increase total margin in both cases.  An increase in rates is, however, correlated with fewer sales.

In [144]:
def tmargin_700():
    tmargin_700_30 = df_margin[(df_margin['fico']==700) 
              & (df_margin['converted to sale']==True)
              & (df_margin['apr']==3.0)]['margin'].mean()

    tmargin_700_40 = df_margin[(df_margin['fico']==700) 
              & (df_margin['converted to sale']==True)
              & (df_margin['apr']==4.0)]['margin'].mean()

    # Create a trace
    trace = go.Scatter(
        y = [tmargin_700_30,tmargin_700_40],
        x = [3.0, 4.0],
        mode = 'lines+markers'
    )

    data = [trace]

    layout = dict(title = 'Ttoal Margin vs APR for 700 FICO',
                  xaxis=dict(title='APR'),
                  yaxis=dict(title='Total Margin') 
                 )

    fig = dict(data=data, layout=layout)
    return fig
fig = tmargin_700()
py.iplot(fig)

In [143]:
def tmargin_600():
    tmargin_700_44 = df_margin[(df_margin['fico']==600) 
              & (df_margin['converted to sale']==True)
              & (df_margin['apr']==4.4)]['margin'].mean()

    tmargin_700_54 = df_margin[(df_margin['fico']==600) 
              & (df_margin['converted to sale']==True)
              & (df_margin['apr']==5.4)]['margin'].mean()


    # Create a trace
    trace = go.Scatter(
        y = [tmargin_700_30,tmargin_700_40,tmargin_700_44,tmargin_700_54],
        x = [4.4, 5.4],
        mode = 'lines+markers'
    )

    data = [trace]

    layout = dict(title = 'Ttoal Margin vs APR for 600 FICO',
                  xaxis=dict(title='APR'),
                  yaxis=dict(title='Total Margin') 
                 )

    fig = dict(data=data, layout=layout)
    return fig
fig = tmargin_600()
py.iplot(fig)

## Q5: 5.	Other than these cut and dry numbers, can you think of other factors Carvana might want to take into account before changing rates? 

•	Customer Satisfaction: Purchasing an car from Carvana is a unique experience where word of mouth can help promote Carvana’s business model.  The method of purchase is so different from other methods of buying a car that there is the possibility of people looking for an excuse NOT to do it.  Evaluating what purchasing customers are saying about the financing could help avoid a possible road block to bringing in other customers if the customer reaction is negative.

•	Brand Sentiment on Social Media:  The reasons for considering social media sentiment are similar to the reasons for considering customer satisfaction.  A baseline of sentiment could be established before the change and evaluated after the change.  